File: packages/flint-wasm/src/number-field-zeta.c
/*
 * WebAssembly storage adapter for the shared number-field zeta residue core.
 *
 * This file owns buffers only.  Factorization and validation live in
 * packages/flint/src/number_field_zeta_core.c, which is also used by the
 * native Node adapter.  One combined uint32 input buffer and one combined
 * uint16 output buffer make the host crossing independent of FLINT layouts.
 *
 * Copyright (C) 2026 Sage.js contributors
 * License: GPL-3.0-only
 */

#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>

#include "number_field_zeta_core.h"

static uint32_t *nf_zeta_input;
static uint16_t *nf_zeta_output;
static uint32_t nf_zeta_degree;
static uint32_t nf_zeta_prime_count;
static size_t nf_zeta_input_word_count;
static size_t nf_zeta_output_word_count;

__attribute__((visibility("default")))
void sagejs_nf_zeta_residue_clear(void)
{
    free(nf_zeta_input);
    free(nf_zeta_output);
    nf_zeta_input = NULL;
    nf_zeta_output = NULL;
    nf_zeta_degree = 0;
    nf_zeta_prime_count = 0;
    nf_zeta_input_word_count = 0;
    nf_zeta_output_word_count = 0;
}

__attribute__((visibility("default")))
int32_t sagejs_nf_zeta_residue_begin(uint32_t degree, uint32_t prime_count)
{
    size_t coefficient_cells;
    size_t factor_cells;

    sagejs_nf_zeta_residue_clear();
    if (degree < 1 || degree > SAGEJS_NF_FACTOR_MAX_DEGREE)
        return SAGEJS_NF_FACTOR_UNSUPPORTED_DEGREE;
    if (prime_count > SAGEJS_NF_FACTOR_MAX_PRIMES)
        return SAGEJS_NF_FACTOR_TOO_MANY_PRIMES;
    if (prime_count == 0)
    {
        nf_zeta_degree = degree;
        return SAGEJS_NF_FACTOR_OK;
    }
    if ((size_t) prime_count > SIZE_MAX / ((size_t) degree + 1U))
        return SAGEJS_NF_FACTOR_SIZE_OVERFLOW;
    coefficient_cells = (size_t) prime_count * ((size_t) degree + 1U);
    factor_cells = (size_t) prime_count * (size_t) degree;
    if (coefficient_cells > SIZE_MAX - (size_t) prime_count ||
        factor_cells > (SIZE_MAX - (size_t) prime_count) / 2U)
        return SAGEJS_NF_FACTOR_SIZE_OVERFLOW;
    nf_zeta_input_word_count = (size_t) prime_count + coefficient_cells;
    nf_zeta_output_word_count = (size_t) prime_count + 2U * factor_cells;
    if (nf_zeta_input_word_count > SIZE_MAX / sizeof(*nf_zeta_input) ||
        nf_zeta_output_word_count > SIZE_MAX / sizeof(*nf_zeta_output))
    {
        sagejs_nf_zeta_residue_clear();
        return SAGEJS_NF_FACTOR_SIZE_OVERFLOW;
    }
    nf_zeta_input = calloc(nf_zeta_input_word_count,
        sizeof(*nf_zeta_input));
    nf_zeta_output = calloc(nf_zeta_output_word_count,
        sizeof(*nf_zeta_output));
    if (nf_zeta_input == NULL || nf_zeta_output == NULL)
    {
        sagejs_nf_zeta_residue_clear();
        return SAGEJS_NF_FACTOR_ALLOCATION_FAILED;
    }
    nf_zeta_degree = degree;
    nf_zeta_prime_count = prime_count;
    return SAGEJS_NF_FACTOR_OK;
}

__attribute__((visibility("default")))
uintptr_t sagejs_nf_zeta_residue_input(void)
{
    return (uintptr_t) nf_zeta_input;
}

__attribute__((visibility("default")))
uint32_t sagejs_nf_zeta_residue_input_words(void)
{
    return (uint32_t) nf_zeta_input_word_count;
}

__attribute__((visibility("default")))
uintptr_t sagejs_nf_zeta_residue_output(void)
{
    return (uintptr_t) nf_zeta_output;
}

__attribute__((visibility("default")))
uint32_t sagejs_nf_zeta_residue_output_words(void)
{
    return (uint32_t) nf_zeta_output_word_count;
}

__attribute__((visibility("default")))
int32_t sagejs_nf_zeta_residue_compute(void)
{
    size_t factor_cells;

    if (nf_zeta_prime_count == 0)
        return nf_zeta_degree == 0
            ? SAGEJS_NF_FACTOR_INVALID_ARGUMENT
            : SAGEJS_NF_FACTOR_OK;
    if (nf_zeta_input == NULL || nf_zeta_output == NULL)
        return SAGEJS_NF_FACTOR_INVALID_ARGUMENT;
    factor_cells = (size_t) nf_zeta_prime_count * (size_t) nf_zeta_degree;
    return sagejs_nf_factor_degrees_residue_batch(
        nf_zeta_output,
        nf_zeta_output + nf_zeta_prime_count,
        nf_zeta_output + nf_zeta_prime_count + factor_cells,
        nf_zeta_input + nf_zeta_prime_count,
        nf_zeta_input,
        nf_zeta_degree,
        nf_zeta_prime_count);
}

File: packages/flint-wasm/number-field-zeta.mjs
const MAX_DEGREE = 64;
const MAX_PRIMES = 65536;
const MAX_WORD_PRIME = 0xffffffffn;

const REQUIRED_EXPORTS = Object.freeze([
  "memory",
  "sagejs_nf_zeta_residue_begin",
  "sagejs_nf_zeta_residue_input",
  "sagejs_nf_zeta_residue_input_words",
  "sagejs_nf_zeta_residue_output",
  "sagejs_nf_zeta_residue_output_words",
  "sagejs_nf_zeta_residue_compute",
  "sagejs_nf_zeta_residue_clear",
]);

function uint32(value) {
  return Number(value) >>> 0;
}

function exactInteger(value, description) {
  if (typeof value === "number" && !Number.isSafeInteger(value)) {
    throw new RangeError(`${description} must be exact; use bigint`);
  }
  try {
    return BigInt(value);
  } catch {
    throw new TypeError(`${description} must be an integer`);
  }
}

function positiveResidue(value, prime) {
  value %= prime;
  return value < 0n ? value + prime : value;
}

function batchError(status) {
  switch (status) {
    case -1:
      return new TypeError("number-field residue batch has invalid buffers");
    case -2:
      return new RangeError("polynomial degree must be between 1 and 64");
    case -3:
      return new RangeError("a factor-degree batch is limited to 65536 primes");
    case -4:
      return new RangeError("number-field residue batch size overflowed");
    case -5:
      return new RangeError("unable to allocate number-field residue buffers");
    default:
      return new Error(`number-field residue batch failed with status ${status}`);
  }
}

/**
 * Create the number-field zeta portion of the FLINT Wasm backend.
 *
 * The public operation deliberately matches the Node addon's packed API, but
 * arbitrary coefficients are reduced in JavaScript before the single packed
 * copy into Wasm. Public factor records are constructed by ordinary Sage.js,
 * not by either host adapter.
 */
export function createNumberFieldZetaBackend(
  instance,
  { recordCapability = () => {} } = {},
) {
  if (!instance || typeof instance !== "object") {
    throw new TypeError("expected an instantiated FLINT WebAssembly module");
  }
  const exports = instance.exports;
  for (const name of REQUIRED_EXPORTS) {
    if (!(name in exports)) {
      throw new Error(`FLINT WebAssembly module is missing export ${name}`);
    }
  }
  const memory = exports.memory;

  function nfFactorDegreesBatch(coefficients, primes) {
    if (!Array.isArray(coefficients)) {
      throw new TypeError("coefficients must be an Array of exact integers");
    }
    const degree = coefficients.length - 1;
    if (degree < 1 || degree > MAX_DEGREE) {
      throw new RangeError("polynomial degree must be between 1 and 64");
    }
    const exactCoefficients = coefficients.map((coefficient, index) =>
      exactInteger(coefficient, `coefficient ${index}`));
    if (exactCoefficients[degree] !== 1n) {
      throw new RangeError("polynomial must be monic");
    }
    if (!(primes instanceof BigUint64Array)) {
      throw new TypeError("primes must be a BigUint64Array");
    }
    const primeCount = primes.length;
    if (primeCount > MAX_PRIMES) {
      throw new RangeError("a factor-degree batch is limited to 65536 primes");
    }
    if (primeCount === 0) {
      return Object.freeze({
        degree,
        primeCount: 0,
        factorCounts: new Uint16Array(),
        exponents: new Uint16Array(),
        degrees: new Uint16Array(),
      });
    }

    const inputWords = primeCount + primeCount * (degree + 1);
    const packedInput = new Uint32Array(inputWords);
    for (let row = 0; row < primeCount; row += 1) {
      const prime = primes[row];
      if (prime > MAX_WORD_PRIME) {
        const error = new RangeError(
          "prime exceeds the WebAssembly word-prime factorization limit",
        );
        error.code = "SAGEJS_WASM_WORD_PRIME_UNAVAILABLE";
        throw error;
      }
      packedInput[row] = Number(prime);
      const offset = primeCount + row * (degree + 1);
      for (let index = 0; index <= degree; index += 1) {
        packedInput[offset + index] = Number(
          positiveResidue(exactCoefficients[index], prime),
        );
      }
    }

    const beginStatus = exports.sagejs_nf_zeta_residue_begin(
      degree,
      primeCount,
    );
    if (beginStatus !== 0) throw batchError(beginStatus);
    try {
      const actualInputWords = uint32(
        exports.sagejs_nf_zeta_residue_input_words(),
      );
      if (actualInputWords !== inputWords) {
        throw new Error("FLINT Wasm returned an inconsistent residue input size");
      }
      const inputPointer = uint32(exports.sagejs_nf_zeta_residue_input());
      new Uint32Array(memory.buffer, inputPointer, inputWords).set(packedInput);
      const status = exports.sagejs_nf_zeta_residue_compute();
      if (status > 0) {
        throw new RangeError(
          `unable to factor the polynomial at supplied prime index ${status - 1}`,
        );
      }
      if (status < 0) throw batchError(status);

      const factorCells = primeCount * degree;
      const outputWords = primeCount + 2 * factorCells;
      const actualOutputWords = uint32(
        exports.sagejs_nf_zeta_residue_output_words(),
      );
      if (actualOutputWords !== outputWords) {
        throw new Error("FLINT Wasm returned an inconsistent residue output size");
      }
      const outputPointer = uint32(exports.sagejs_nf_zeta_residue_output());
      // One copied egress keeps returned views valid across memory growth and
      // permits the C-owned buffers to be cleared immediately.
      const output = new Uint16Array(
        memory.buffer,
        outputPointer,
        outputWords,
      ).slice();
      recordCapability(
        "napi:@sagemath/sagejs-flint:nfFactorDegreesBatch",
        "receipt-backed-wasm-artifact",
        {
          executionTarget: "wasm-artifact",
          ingressBytes: packedInput.byteLength,
          egressBytes: output.byteLength,
        },
      );
      return Object.freeze({
        degree,
        primeCount,
        factorCounts: output.subarray(0, primeCount),
        exponents: output.subarray(primeCount, primeCount + factorCells),
        degrees: output.subarray(primeCount + factorCells),
      });
    } finally {
      exports.sagejs_nf_zeta_residue_clear();
    }
  }

  return Object.freeze({
    nfFactorDegreesBatch,
    nfFactorDegreesBatchMaxPrime: MAX_WORD_PRIME,
  });
}

export const numberFieldZetaWasmExports = REQUIRED_EXPORTS;
